### ISSUE TITLE
Question about the datasets

### ISSUE THREAD

**Issue by liting1018** [yangzhou321/VQBasedRepresentationLearning#3]
Thanks for sharing your nice work！ I would like to know how the datasets was prepared?

**Comment by yangzhou321** [comment]
1、For model training, you can download the ImageNet-1k dataset from [official](https://image-net.org/download.php) or [kaggle](https://www.kaggle.com/competitions/imagenet-object-localization-challenge/data) and change the "--root_path" in line 21 in [train_sa.py](https://github.com/yangzhou321/VQSA/blob/main/train_sa.py) to your data path
2、For the test dataset ImageNet-C you can download it from this [website](https://zenodo.org/records/2235448)(it's a validation set): or use the python library "imagenet-c" or "imagecorruption" to generate the corrupt image (for training and validation).  (see [imagecorruptions](https://github.com/bethgelab/imagecorruptions) for detail.) For ImageNet-A and -R, please refer to [robusta](https://github.com/bethgelab/robustness). And please also change the corresponding "img_path" in line 78 in [train_sa.py](https://github.com/yangzhou321/VQSA/blob/main/train_sa.py) and line 69 in [test_sa.py](https://github.com/yangzhou321/VQSA/blob/main/test_sa.py)

**Comment by liting1018** [comment]
非常感谢你的回复！那么我还有个疑问，训练阶段的输入只包括清晰的样本吗？但文章中的图3展示的输入包括了清晰和对应的模糊的图片，这里我理解有点歧义。如果训练阶段输入只包括清晰的图片，用量化操作就是为了提取与图像质量无关的特征，以适应测试的时候数据集是模糊的图片是嘛？

**Comment by liting1018** [comment]
我看训练阶段代码中is_paired都为false，说明清晰的图和模糊的图没有拼起来送到网络中，跟文章中的结构图不一致呢。另外，加载相应的数据集，你分享的代码是不是可以直接跑起来哇，我想调试看看。我理解的可能有点问题，麻烦解答哈疑惑呢，谢谢啦~

**Comment by yangzhou321** [comment]
1、训练阶段的输入包括清晰的样本和仿真的退化样本，通过[imagenet_dataset.py](https://github.com/yangzhou321/VQSA/blob/main/imagenet_dataset.py)中的get_CrpNameAndSevLevel函数生成各种退化。在大量的清晰+退化图像中通过量化，学习图像不变的表示特征，即“图像质量无关的特征”
2、is_paired为false是代表数据以非成对的方式进行训练的哈，因为我之前的工作是成对图像作蒸馏，所以沿用了类似的框架。文章图想表达的就是说，我的训练集包含了干净的图像和仿真的退化图像，但不是成对的方式训练。代码只要数据正确，库“imagecorruption”安装好，应该可以跑通。我这边是已经测试过可以跑通。
3、有任何问题欢迎继续讨论呀。

**Comment by liting1018** [comment]
好的好的，太感谢啦~

**Comment by liting1018** [comment]
我部署了数据集已经调通代码了，但还存在以下问题：
（1）关于论文中的公式（6）不是很理解，通过MAP从z恢复z cat是代码中的那个直通估计quantized = x + (quantized - x).detach()嘛？

（2）另外，对于直接设置系数为one-hot vector，我也不知道代码是如何体现的，代码的codebook就是用embedding构建的，利用索引去查找相似的值。
期望得到你的解答呐，感谢！

**Comment by yangzhou321** [comment]
(1)、不是的，quantized = x + (quantized - x).detach()这一仅仅是梯度的Straight-Through Estimator，也就是你说的直通估计，它只是用来优化codebook参数的（很显然这里将x的梯度直接传给了从codebook中量化所得的quantized，这样反向传播就会优化到codebook中的参数，不然的话公式（3）量化的过程是不可导的）。和估计z hat无关。估计z hat的过程在本方法的实验中其实就是量化的过程，通过量化丢弃噪声等干扰，保留本质用于分类的特征。公式（6）是经典的稀疏表示求解公式，我写在论文中是想说，受到稀疏表示字典学习可以对信号去噪这一启发来的。而我们难以用传统的方式来求解公式（6），所以用VQ量化的方式。他们其实是等价的。
(2)、标准的的稀疏表示中$\alpha$表示系数是一个稀疏向量，意思是从字典中找很少的几个组元来表示原始信号，尽管稀疏，但并不是one hot（one hot就是只有一个为1，意思就是只取一个），如果想让$\alpha$成为one hot，意思就是要从字典中仅仅寻找一个组元来作为表示，其实就是相当于从codebook中找最相似的那一个特征。然而向量量化（VQ）的理论中，正好就是这样子。总结来说：设置表示系数$\alpha$为one hot，等价于使用标准的向量量化。当你使用公式（3）即代码”train_sa.py“的216行get_code_indices函数选择最相似的codebook组元来表示的时候，就相当于完成了设置$\alpha$为one hot。

以上，不知道我解释的你是否能明白哈。你还可以参考文章中提到的VQVAE论文以及2.2节的内容及相关文献，有助于更好的理解本文思路。

**Comment by liting1018** [comment]
> 1、训练阶段的输入包括清晰的样本和仿真的退化样本，通过[imagenet_dataset.py](https://github.com/yangzhou321/VQSA/blob/main/imagenet_dataset.py)中的get_CrpNameAndSevLevel函数生成各种退化。在大量的清晰+退化图像中通过量化，学习图像不变的表示特征，即“图像质量无关的特征” 2、is_paired为false是代表数据以非成对的方式进行训练的哈，因为我之前的工作是成对图像作蒸馏，所以沿用了类似的框架。文章图想表达的就是说，我的训练集包含了干净的图像和仿真的退化图像，但不是成对的方式训练。代码只要数据正确，库“imagecorruption”安装好，应该可以跑通。我这边是已经测试过可以跑通。 3、有任何问题欢迎继续讨论呀。

看了ImageNet_DataSet这里，最开始输入是imagenet里的干净图像，通过get_CrpNameAndSevLevel得到退化图像distort_img，如果是不配对的情况，最终送进网络的只有distort_img啊，如果是不配对的情况，才是将img和distort_img拼接一起送进去网络了。对应代码是这里
if self.is_pair:
            img = torch.cat([img, distort_img], dim=0)
            return img, target, index  # train
        else:
            return distort_img, target, index  # test corrupt
     能麻烦帮我解答哈嘛，谢谢~

**Comment by liting1018** [comment]
> (1)、不是的，quantized = x + (quantized - x).detach()这一仅仅是梯度的Straight-Through Estimator，也就是你说的直通估计，它只是用来优化codebook参数的（很显然这里将x的梯度直接传给了从codebook中量化所得的quantized，这样反向传播就会优化到codebook中的参数，不然的话公式（3）量化的过程是不可导的）。和估计z hat无关。估计z hat的过程在本方法的实验中其实就是量化的过程，通过量化丢弃噪声等干扰，保留本质用于分类的特征。公式（6）是经典的稀疏表示求解公式，我写在论文中是想说，受到稀疏表示字典学习可以对信号去噪这一启发来的。而我们难以用传统的方式来求解公式（6），所以用VQ量化的方式。他们其实是等价的。 (2)、标准的的稀疏表示中$\alpha$表示系数是一个稀疏向量，意思是从字典中找很少的几个组元来表示原始信号，尽管稀疏，但并不是one hot（one hot就是只有一个为1，意思就是只取一个），如果想让$\alpha$成为one hot，意思就是要从字典中仅仅寻找一个组元来作为表示，其实就是相当于从codebook中找最相似的那一个特征。然而向量量化（VQ）的理论中，正好就是这样子。总结来说：设置表示系数$\alpha$为one hot，等价于使用标准的向量量化。当你使用公式（3）即代码”train_sa.py“的216行get_code_indices函数选择最相似的codebook组元来表示的时候，就相当于完成了设置$\alpha$为one hot。
> 
> 以上，不知道我解释的你是否能明白哈。你还可以参考文章中提到的VQVAE论文以及2.2节的内容及相关文献，有助于更好的理解本文思路。

感谢你的解释，了解了相关工作，明白辣！

**Comment by yangzhou321** [comment]
> > 1、训练阶段的输入包括清晰的样本和仿真的退化样本，通过[imagenet_dataset.py](https://github.com/yangzhou321/VQSA/blob/main/imagenet_dataset.py)中的get_CrpNameAndSevLevel函数生成各种退化。在大量的清晰+退化图像中通过量化，学习图像不变的表示特征，即“图像质量无关的特征” 2、is_paired为false是代表数据以非成对的方式进行训练的哈，因为我之前的工作是成对图像作蒸馏，所以沿用了类似的框架。文章图想表达的就是说，我的训练集包含了干净的图像和仿真的退化图像，但不是成对的方式训练。代码只要数据正确，库“imagecorruption”安装好，应该可以跑通。我这边是已经测试过可以跑通。 3、有任何问题欢迎继续讨论呀。
> 
> 看了ImageNet_DataSet这里，最开始输入是imagenet里的干净图像，通过get_CrpNameAndSevLevel得到退化图像distort_img，如果是不配对的情况，最终送进网络的只有distort_img啊，如果是不配对的情况，才是将img和distort_img拼接一起送进去网络了。对应代码是这里 if self.is_pair: img = torch.cat([img, distort_img], dim=0) return img, target, index # train else: return distort_img, target, index # test corrupt 能麻烦帮我解答哈嘛，谢谢~

你仔细看下"imagenet_dataset.py"的第51和80行，如果sample出的crp_idx == -1，会直接返回干净的图像哈

**Comment by liting1018** [comment]
> > > 1、训练阶段的输入包括清晰的样本和仿真的退化样本，通过[imagenet_dataset.py](https://github.com/yangzhou321/VQSA/blob/main/imagenet_dataset.py)中的get_CrpNameAndSevLevel函数生成各种退化。在大量的清晰+退化图像中通过量化，学习图像不变的表示特征，即“图像质量无关的特征” 2、is_paired为false是代表数据以非成对的方式进行训练的哈，因为我之前的工作是成对图像作蒸馏，所以沿用了类似的框架。文章图想表达的就是说，我的训练集包含了干净的图像和仿真的退化图像，但不是成对的方式训练。代码只要数据正确，库“imagecorruption”安装好，应该可以跑通。我这边是已经测试过可以跑通。 3、有任何问题欢迎继续讨论呀。
> > 
> > 
> > 看了ImageNet_DataSet这里，最开始输入是imagenet里的干净图像，通过get_CrpNameAndSevLevel得到退化图像distort_img，如果是不配对的情况，最终送进网络的只有distort_img啊，如果是不配对的情况，才是将img和distort_img拼接一起送进去网络了。对应代码是这里 if self.is_pair: img = torch.cat([img, distort_img], dim=0) return img, target, index # train else: return distort_img, target, index # test corrupt 能麻烦帮我解答哈嘛，谢谢~
> 
> 你仔细看下"imagenet_dataset.py"的第51和80行，如果sample出的crp_idx == -1，会直接返回干净的图像哈

看到啦，谢谢~
